src/coreclr/src/jit/morph.cpp

```diff
@@ -7298,306 +7283,724 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
+        // We have removed all statements after this statement and set the
+        // current statement expression to just be the call. However, we may
+        // still be morphing a parent node (such as a return) which will expect
+        // us to return something of a valid type. We thus return a placeholder
+        // node here that will not be used for anything since we have effectively
+        // removed the parent node from the tree already.
+        if (origCallType != TYP_VOID && info.compRetType != TYP_VOID)
+        {
+            var_types nodeTy = origCallType;
+#ifdef FEATURE_HFA
+            // Return a dummy node, as the return is already removed.
+            if (nodeTy == TYP_STRUCT)
+            {
+                // This is a HFA, use float 0.
+                nodeTy = TYP_FLOAT;
+            }
+#elif defined(UNIX_AMD64_ABI)
+            // Return a dummy node, as the return is already removed.
+            if (varTypeIsStruct(nodeTy))
+            {
+                // This is a register-returned struct. Return a 0.
+                // The actual return registers are hacked in lower and the register allocator.
+                nodeTy = TYP_INT;
+            }
 #endif
 #ifdef FEATURE_SIMD
-        // Return a dummy node, as the return is already removed.
-        if (varTypeIsSIMD(nodeTy))
-        {
-            nodeTy = TYP_DOUBLE;
-        }
+            // Return a dummy node, as the return is already removed.
+            if (varTypeIsSIMD(nodeTy))
+            {
+                nodeTy = TYP_DOUBLE;
+            }
 #endif
-        result = gtNewZeroConNode(genActualType(nodeTy));
-        result = fgMorphTree(result);
+            result = gtNewZeroConNode(genActualType(nodeTy));
+            result = fgMorphTree(result);
+        }
     }
 
     return result;
 }
 
 /*****************************************************************************
  *
- *  Transform the given GT_CALL tree for tail call code generation.
+ * Transform the given GT_CALL tree for tail call code generation.
+ * This transforms
+ *   CALL
+ *       {callTarget}
+ *       {this}
+ *       {args}
+ * to
+ *   COMMA
+ *     CALL StoreArgsStub
+ *       {callTarget} // depending on flags provided by the runtime
+ *       {this}       // as a regular arg
+ *       {args}
+ *     COMMA
+ *       CALL Dispatcher
+ *         ADDR ReturnAddress
+ *         CallTargetStub
+ *         ADDR ReturnValue
+ *       LCL ReturnValue
+ * whenever the call node returns a value. If the call node does not return a
+ * value the last comma will not be there.
  */
-void Compiler::fgMorphTailCallViaHelper(GenTreeCall* call, void* pfnCopyArgs)
+GenTree* Compiler::fgMorphTailCallViaHelpers(GenTreeCall* call, CORINFO_TAILCALL_HELPERS& help)
 {
-#if defined(_TARGET_UNIX_)
-    noway_assert(!"Slow tail calls not supported on non-Windows platforms.");
-#endif
-
-    JITDUMP("fgMorphTailCallViaHelper (before):\n");
+    JITDUMP("fgMorphTailCallViaHelpers (before):\n");
```

Comment:
This method in general makes interestingly scary changes to the argument list. I think in general this method should have many asserts. Any time the arg list changes there should be an assert that the rest of the arglist is unaffected. Also there should be entry and exit asserts checking that the arglist is valid.

Reply:
Do you have suggestions on how to check those conditions?

Reply:
Didn't think that far ahead sorry, I spoke with @BruceForstall and he and I didn't have a clear idea on how to do this today. Would you mind adding `TODO: assert check this arg is the correct arg type to remove`

Reply:
We can address this post-merge.